build-scripts/graal.js

```diff
@@ -41,86 +41,76 @@ process.on('unhandledRejection', error => {
   process.exit(1);
 });
 
-/**
- * Simple wrapper for the NodeJS spawn command to use promises
- *
- * @param {string} command
- * @param {Object=} spawnOpts
- * @return {!Promise<undefined>}
- */
-function runCommand(command, spawnOpts = {}) {
-  // log command being executed to stdout for debugging
-  process.stdout.write(`${command}\n`);
-  return new Promise((resolve, reject) => {
-    // TODO(ChadKillingsworth): Not really safe in general, since this could split in the middle of quoted strings.
-    // This is good enough for the purposes of this script.
-    const commandParts = command.split(/\s+/);
-    // child process should inherit stdin/out/err from this process unless spawnOpts says otherwise
-    const opts = Object.assign({}, {
-      stdio: 'inherit'
-    }, spawnOpts);
-    const externalProcess = spawn(commandParts[0], commandParts.slice(1), opts);
-    externalProcess.on('error', err => {
-      process.stderr.write(`${err.toString()}\n`);
-      reject(err);
-    });
-    externalProcess.on('close', exitCode => {
-      if (exitCode != 0) {
-        process.stderr.write(`non zero exit code ${exitCode}\n`);
-        process.exit(1);
-        reject();
-      }
-      resolve();
-    });
-  });
-}
-
 // Build graal from source
 if (!fs.existsSync(TEMP_PATH)) {
   fs.mkdirSync(TEMP_PATH);
 }
 
-const NATIVE_IMAGE_BUILD_ARGS = [
-  '-H:+JNI',
-  '--no-server',
+const NATIVE_IMAGE_BUILD_ARGS = [ '-H:+JNI'].concat(GRAAL_OS !== 'windows' ? ['--no-server'] : []).concat([
   '-H:+ReportUnsupportedElementsAtRuntime',
   '-H:IncludeResourceBundles=com.google.javascript.rhino.Messages',
   '-H:IncludeResourceBundles=org.kohsuke.args4j.Messages',
   '-H:IncludeResourceBundles=org.kohsuke.args4j.spi.Messages',
   '-H:IncludeResourceBundles=com.google.javascript.jscomp.parsing.ParserConfig',
   `-H:ReflectionConfigurationFiles=${path.resolve(__dirname, 'reflection-config.json')}`,
-  '-H:IncludeResources=(externs.zip)|(.*(js|txt))',
+  '-H:IncludeResources=(externs.zip)|(.*(js|txt))'.replace(/\|/g, () => {
+    if (GRAAL_OS === 'windows') {
+      return '^^^|';
```

Comment:
OK, I give up.  Why these three ^'s on windows?

Reply:
Crazy window shell escaping. You escape with a carret and you have to double escape one. You would not believe how long this took to find. Guess I need to add a comment.